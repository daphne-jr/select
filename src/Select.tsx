import * as React from 'react';
import KeyCode from 'rc-util/lib/KeyCode';
import classNames from 'classnames';
import Selector, { RefSelectorProps } from './Selector';
import SelectTrigger from './SelectTrigger';
import { SelectContext } from './Context';
import { RenderNode, OptionsType as SelectOptionsType } from './interface';
import {
  RawValueType,
  LabelValueType,
  ValueType,
  GetLabeledValue,
  FilterOptions,
  FilterFunc,
} from './interface/generator';
import SelectOptionList, { OptionListProps, RefProps } from './OptionList';
import Option from './Option';
import OptGroup from './OptGroup';
import { convertChildrenToData as convertSelectChildrenToData } from './utils/legacyUtil';
import {
  getLabeledValue as getSelectLabeledValue,
  filterOptions as selectDefaultFilterOptions,
} from './utils/valueUtil';
import { toInnerValue, toOuterValues } from './utils/commonUtil';
import TransBtn from './TransBtn';

/**
 * To match accessibility requirement, we always provide an input in the component.
 * Other element will not set `tabIndex` to avoid `onBlur` sequence problem.
 * For focused select, we set `aria-live="polite"` to update the accessibility content.
 *
 * ref:
 * - keyboard: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role#Keyboard_interactions
 */

export interface SelectProps<OptionsType> {
  prefixCls?: string;
  id?: string;
  className?: string;

  // Options
  options?: OptionsType;
  children?: React.ReactNode;
  mode?: 'multiple' | 'tags' | 'combobox';

  // Value
  value?: ValueType;
  defaultValue?: ValueType;
  labelInValue?: boolean;

  // Search
  searchValue?: string;
  optionFilterProp?: string;
  /**
   * In Select, `false` means do nothing.
   * In TreeSelect, `false` will highlight match item.
   * It's by design.
   */
  filterOption?: boolean | FilterFunc;
  autoClearSearchValue?: boolean;
  onSearch?: (value: string) => void;

  // Icons
  clearIcon?: React.ReactNode;

  // Others
  autoFocus?: boolean;
  allowClear?: boolean;

  // Events
  onKeyUp?: React.KeyboardEventHandler<HTMLDivElement>;
  onKeyDown?: React.KeyboardEventHandler<HTMLDivElement>;

  // Legacy
  defaultActiveFirstOption?: boolean;
  combobox?: boolean;

  showSearch?: boolean;
  disabled?: boolean;
  style?: React.CSSProperties;
  showArrow?: boolean;
  openClassName?: string;
  transitionName?: string;
  optionLabelProp?: string;
  animation?: string;
  choiceTransitionName?: string;
  open?: boolean;
  defaultOpen?: boolean;
  inputValue?: string;
  onClick?: React.MouseEventHandler;
  onChange?: (value: ValueType, option: JSX.Element | JSX.Element[]) => void;
  onBlur?: React.FocusEventHandler<HTMLInputElement>;
  onFocus?: React.FocusEventHandler<HTMLInputElement>;
  onSelect?: (value: ValueType, option: JSX.Element | JSX.Element[]) => void;
  onDropdownVisibleChange?: (open: boolean | undefined) => void;
  onPopupScroll?: React.UIEventHandler<HTMLDivElement>;
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>;
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
  onInputKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  placeholder?: string;
  onDeselect?: (value: ValueType, option: JSX.Element | JSX.Element[]) => void;
  loading?: boolean;
  firstActiveValue?: ValueType;
  dropdownStyle?: React.CSSProperties;
  maxTagTextLength?: number;
  maxTagCount?: number;
  maxTagPlaceholder?: RenderNode;
  tokenSeparators?: string[];
  getInputElement?: () => JSX.Element;
  showAction?: string[];
  inputIcon?: React.ReactNode;
  removeIcon?: React.ReactNode;
  menuItemSelectedIcon?: RenderNode;
  getPopupContainer?: RenderNode;
  dropdownRender?: (menu: any) => JSX.Element;
  backfill?: boolean;
  dropdownAlign?: any;
  dropdownClassName?: string;
  dropdownMatchSelectWidth?: boolean;
  dropdownMenuStyle?: React.CSSProperties;
  notFoundContent?: string | false;
  tabIndex?: number;
}

export interface GenerateConfig<OptionsType, StaticProps> {
  components: {
    optionList: React.ForwardRefExoticComponent<
      React.PropsWithoutRef<Omit<OptionListProps, 'options'> & { options: OptionsType }> &
        React.RefAttributes<RefProps>
    >;
  };
  staticProps?: StaticProps;
  /** Convert jsx tree into `OptionsType` */
  convertChildrenToData: (children: React.ReactNode) => OptionsType;
  /** Convert single raw value into { label, value } format. Will be called by each value */
  getLabeledValue: GetLabeledValue<OptionsType>;
  filterOptions: FilterOptions<OptionsType>;
}

/**
 * This function is in internal usage.
 * Do not use it in your prod env since we may refactor this.
 */
export function generateSelector<OptionsType, StaticProps>(
  config: GenerateConfig<OptionsType, StaticProps>,
): React.ComponentType<SelectProps<OptionsType>> {
  /** Used for accessibility id generate */
  let uuid = 0;

  const {
    components: { optionList: OptionList },
    staticProps,
    convertChildrenToData,
    getLabeledValue,
    filterOptions,
  } = config;

  type SelectComponent = React.FC<SelectProps<OptionsType>> & StaticProps;

  const Select: React.FC<SelectProps<OptionsType>> = props => {
    const {
      prefixCls = 'rc-select',

      id,
      mode,
      value,
      defaultValue,

      // Search related
      showSearch,
      searchValue,
      filterOption,
      optionFilterProp,
      autoClearSearchValue = true,
      onSearch,

      // Icons
      allowClear,
      clearIcon,

      disabled,
      labelInValue,
      className,
      open,
      defaultOpen,
      options,
      children,
      onClick,

      onFocus,
      onBlur,
      onKeyUp,
      onKeyDown,
      onMouseDown,

      onChange,
      onSelect,
      onDeselect,

      ...domProps
    } = props;

    const selectorRef = React.useRef<RefSelectorProps>(null);
    const listRef = React.useRef<RefProps>(null);

    // Inner id for accessibility usage. Only work in client side
    const [innerId, setInnerId] = React.useState<string>();
    React.useEffect(() => {
      setInnerId(`rc_select_${uuid}`);
      uuid += 1;
    }, []);
    const mergedId = id || innerId;

    // ============================= Option =============================
    const [innerSearchValue, setInnerSearchValue] = React.useState('');
    const mergedSearchValue = searchValue !== undefined ? searchValue : innerSearchValue;

    const baseOptions = React.useMemo<OptionsType>((): OptionsType => {
      if (options !== undefined) {
        return options;
      }

      return convertChildrenToData(children);
    }, [options, children]);

    const mergedOptions = React.useMemo<OptionsType>(() => {
      if (!mergedSearchValue) {
        return baseOptions;
      }
      return filterOptions(mergedSearchValue, baseOptions, props);
    }, [baseOptions, mergedSearchValue]);

    const triggerSearch = (searchText: string) => {
      setInnerSearchValue(searchText);

      if (onSearch && mergedSearchValue !== searchText) {
        onSearch(searchText);
      }
    };

    // ============================= Value ==============================
    const [innerValue, setInnerValue] = React.useState<ValueType>(value || defaultValue);
    const baseValue = value !== undefined && value !== null ? value : innerValue;

    /** Unique raw values */
    const mergedRawValue = React.useMemo<RawValueType[]>(
      () => toInnerValue(baseValue, { labelInValue }),
      [baseValue],
    );
    /** We cache a set of raw values to speed up check */
    const rawValues = React.useMemo<Set<RawValueType>>(() => new Set(mergedRawValue), [
      mergedRawValue,
    ]);

    const displayValues = React.useMemo<LabelValueType[]>(
      () =>
        mergedRawValue.map((val: RawValueType) => getLabeledValue(val, mergedOptions, baseValue)),
      [baseValue],
    );

    const isMultiple = mode === 'tags' || mode === 'multiple';

    const triggerChange = React.useCallback(
      (values: RawValueType[] | LabelValueType[]) => {
        const outValue: ValueType = isMultiple ? values : values[0];
        // Skip trigger if prev & current value is both empty
        if (onChange && (mergedRawValue.length !== 0 || values.length !== 0)) {
          // TODO: handle this
          onChange(outValue, null);
        }

        setInnerValue(outValue);
      },
      [onChange, isMultiple, mergedRawValue],
    );

    const onInternalSelect = (newValue: RawValueType, { selected }: { selected: boolean }) => {
      let newRawValue: Set<RawValueType>;

      if (isMultiple) {
        newRawValue = new Set(mergedRawValue);
        if (selected) {
          newRawValue.add(newValue);
        } else {
          newRawValue.delete(newValue);
        }
      } else {
        newRawValue = new Set();
        newRawValue.add(newValue);
      }

      // Multiple always trigger change and single should change if value changed
      if (isMultiple || (!isMultiple && Array.from(mergedRawValue)[0] !== newValue)) {
        const outValue = toOuterValues<OptionsType>(Array.from(newRawValue), {
          labelInValue,
          options: mergedOptions,
          getLabeledValue,
          prevValue: baseValue,
        });

        triggerChange(outValue);
      }

      // Trigger `onSelect`
      const selectValue = labelInValue
        ? getLabeledValue(newValue, mergedOptions, baseValue)
        : newValue;

      // TODO: second param
      // Single mode always trigger `onSelect`
      if ((!isMultiple || selected) && onSelect) {
        onSelect(selectValue, null);
      } else if (!selected && onDeselect) {
        onDeselect(selectValue, null);
      }

      // Clean search value if single or configured
      if (!isMultiple || autoClearSearchValue) {
        setInnerSearchValue('');
      }
    };

    // ============================== Open ==============================
    const [innerOpen, setInnerOpen] = React.useState<boolean>(defaultOpen);
    const mergeOpen: boolean = open !== undefined ? open : innerOpen;

    const onToggleOpen = React.useCallback<(open?: boolean) => void>(
      (newOpen?: boolean) => {
        const nextOpen = newOpen !== undefined ? newOpen : !mergeOpen;
        if (innerOpen !== nextOpen) {
          setInnerOpen(nextOpen);
        }
      },
      [innerOpen],
    );

    // KeyDown
    const onInternalKeyDown: React.KeyboardEventHandler<HTMLDivElement> = (event, ...rest) => {
      const { which } = event;

      // We only manage open state here, close logic should handle by list component
      if (!mergeOpen && which === KeyCode.ENTER) {
        onToggleOpen(true);
      }

      if (mergeOpen && listRef.current) {
        listRef.current.onKeyDown(event, ...rest);
      }

      if (onKeyDown) {
        onKeyDown(event, ...rest);
      }
    };

    // KeyUp
    const onInternalKeyUp: React.KeyboardEventHandler<HTMLDivElement> = (event, ...rest) => {
      if (mergeOpen && listRef.current) {
        listRef.current.onKeyUp(event, ...rest);
      }

      if (onKeyUp) {
        onKeyUp(event, ...rest);
      }
    };

    // ========================== Focus / Blur ==========================
    const [focused, setFocused] = React.useState(false);

    const onInternalFocus = React.useCallback<React.FocusEventHandler<HTMLInputElement>>(
      (...args) => {
        setFocused(true);
        if (onFocus) {
          onFocus(...args);
        }
      },
      [onFocus],
    );
    const onInternalBlur = React.useCallback<React.FocusEventHandler<HTMLInputElement>>(
      (...args) => {
        setFocused(false);

        // TODO: cancel comment this
        // setInnerOpen(false);
        if (onBlur) {
          onBlur(...args);
        }
      },
      [onBlur],
    );

    const onInternalMouseDown: React.MouseEventHandler<HTMLDivElement> = (...args) => {
      // Not lose focus if have
      if (focused && args[0].target !== selectorRef.current.getInputElement()) {
        args[0].preventDefault();
      }

      if (onMouseDown) {
        onMouseDown(...args);
      }
    };

    // ============================= Popup ==============================
    const popupNode = (
      <OptionList
        ref={listRef}
        prefixCls={prefixCls}
        id={mergedId}
        options={mergedOptions}
        multiple={isMultiple}
        values={rawValues}
        onSelect={onInternalSelect}
        onToggleOpen={onToggleOpen}
      />
    );

    // ============================= Clear ==============================
    let clearNode: React.ReactNode;
    const onClearClick: React.MouseEventHandler<HTMLSpanElement> = () => {
      triggerChange([]);
      triggerSearch('');
    };

    if (allowClear && (mergedRawValue.length || mergedSearchValue)) {
      clearNode = (
        <TransBtn
          className={`${prefixCls}-selection-clear`}
          onClick={onClearClick}
          customizeIcon={clearIcon}
        />
      );
    }

    // ============================= Render =============================
    const mergedClassName = classNames(prefixCls, className, {
      [`${prefixCls}-focused`]: focused,
      [`${prefixCls}-multiple`]: isMultiple,
      [`${prefixCls}-allow-clear`]: allowClear,
    });

    return (
      <SelectContext.Provider value={{ prefixCls }}>
        <div
          className={mergedClassName}
          {...domProps}
          onMouseDown={onInternalMouseDown}
          onKeyDown={onInternalKeyDown}
          onKeyUp={onInternalKeyUp}
        >
          {mergeOpen && (
            <span
              style={{ width: 0, height: 0, display: 'flex', overflow: 'hidden', opacity: 0 }}
              aria-live="assertive"
            >
              TODO: Options
            </span>
          )}
          <SelectTrigger
            disabled={disabled}
            prefixCls={prefixCls}
            visible={mergeOpen}
            popupElement={popupNode}
          >
            <Selector
              {...props}
              ref={selectorRef}
              id={mergedId}
              multiple={isMultiple}
              values={displayValues}
              open={mergeOpen}
              onToggleOpen={onToggleOpen}
              onFocus={onInternalFocus}
              onBlur={onInternalBlur}
              searchValue={mergedSearchValue}
              onSearch={triggerSearch}
            />
          </SelectTrigger>

          {clearNode}
        </div>
      </SelectContext.Provider>
    );
  };

  Select.defaultProps = {
    optionFilterProp: 'value',
  };

  // Inject static props
  if (staticProps) {
    Object.keys(staticProps).forEach(prop => {
      Select[prop] = staticProps[prop];
    });
  }

  return Select as SelectComponent;
}

interface SelectStaticProps {
  Option: typeof Option;
  OptGroup: typeof OptGroup;
}

export default generateSelector<SelectOptionsType, SelectStaticProps>({
  components: {
    optionList: SelectOptionList,
  },
  staticProps: {
    Option,
    OptGroup,
  },
  convertChildrenToData: convertSelectChildrenToData,
  getLabeledValue: getSelectLabeledValue,
  filterOptions: selectDefaultFilterOptions,
});
