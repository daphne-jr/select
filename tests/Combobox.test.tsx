import { mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { resetWarned } from 'rc-util/lib/warning';
import Select, { OptGroup, Option, SelectProps } from '../src';
import focusTest from './shared/focusTest';
import blurTest from './shared/blurTest';
import keyDownTest from './shared/keyDownTest';
import inputFilterTest from './shared/inputFilterTest';
import openControlledTest from './shared/openControlledTest';
import {
  expectOpen,
  toggleOpen,
  selectItem,
  findSelection,
  injectRunAllTimers,
} from './utils/common';
import allowClearTest from './shared/allowClearTest';
import throwOptionValue from './shared/throwOptionValue';

describe('Combobox', () => {
  injectRunAllTimers(jest);

  allowClearTest('combobox', '2333');
  throwOptionValue('combobox');
  focusTest('combobox');
  keyDownTest('combobox');
  openControlledTest('combobox');

  it('renders correctly', () => {
    const wrapper = mount(
      <Select mode="combobox" placeholder="Search">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('set inputValue based on value', () => {
    const wrapper = mount(
      <Select mode="combobox" value="1">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

    expect(wrapper.find('input').props().value).toEqual('1');
  });

  it('placeholder', () => {
    const wrapper = mount(
      <Select mode="combobox" placeholder="placeholder">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

      expect(wrapper.find('input').props().value).toBe('');
      expect(wrapper.find('.rc-select-selection__placeholder').text()).toBe('placeholder');
    //   expect(wrapper.find('.rc-select-selection__placeholder').prop('style')).toHaveProperty(
    //     'display',
    //     'block',
    //   );
    //   wrapper.find('input').simulate('change', { target: { value: '1' } });
    //   expect(
    //     wrapper
    //       .update()
    //       .find('.rc-select-selection__placeholder')
    //       .prop('style'),
    //   ).toHaveProperty('display', 'none');
    //   expect(wrapper.state().inputValue).toBe('1');
  });

  // it('fire change event immediately when user inputing', () => {
  //   const handleChange = jest.fn();
  //   const wrapper = mount(
  //     <Select mode="combobox" onChange={handleChange}>
  //       <Option value="11">11</Option>
  //       <Option value="22">22</Option>
  //     </Select>,
  //   );

  //   wrapper.find('input').simulate('change', { target: { value: '1' } });

  //   expect(handleChange).toBeCalledWith(
  //     '1',
  //     <Option key="1" value="1">
  //       1
  //     </Option>,
  //   );
  // });

  // it('set inputValue when user select a option', () => {
  //   const wrapper = mount<Select>(
  //     <Select mode="combobox">
  //       <Option value="1">1</Option>
  //       <Option value="2">2</Option>
  //     </Select>,
  //   );

  //   wrapper.find('.rc-select').simulate('click');
  //   wrapper
  //     .find('MenuItem')
  //     .first()
  //     .simulate('click');
  //   expect(wrapper.state().inputValue).toBe('1');
  // });

  // describe('input value', () => {
  //   const createSelect = props =>
  //     mount(
  //       <Select mode="combobox" optionLabelProp="children" {...props}>
  //         <Option value="1">One</Option>
  //         <Option value="2">Two</Option>
  //       </Select>,
  //     );

  //   describe('labelInValue is false', () => {
  //     it('displays correct input value for defaultValue', () => {
  //       const wrapper = createSelect({
  //         defaultValue: '1',
  //       });
  //       expect(wrapper.find('input').props().value).toBe('One');
  //     });

  //     it('displays correct input value for value', () => {
  //       const wrapper = createSelect({
  //         value: '1',
  //       });
  //       expect(wrapper.find('input').props().value).toBe('One');
  //     });
  //   });

  //   describe('labelInValue is true', () => {
  //     it('displays correct input value for defaultValue', () => {
  //       const wrapper = createSelect({
  //         labelInValue: true,
  //         defaultValue: { key: '1', label: 'One' },
  //       });
  //       expect(wrapper.find('input').props().value).toBe('One');
  //     });

  //     it('displays correct input value for value', () => {
  //       const wrapper = createSelect({
  //         labelInValue: true,
  //         value: { key: '1', label: 'One' },
  //       });
  //       expect(wrapper.find('input').props().value).toBe('One');
  //     });

  //     it('displays correct input value when value changes', () => {
  //       const wrapper = createSelect({
  //         labelInValue: true,
  //         value: { key: '' },
  //       });
  //       wrapper.setProps({ value: { key: '1', label: 'One' } });
  //       expect(wrapper.find('input').props().value).toBe('One');
  //     });
  //   });

  //   describe('hidden when filtered options is empty', () => {
  //     // https://github.com/ant-design/ant-design/issues/3958
  //     it('should popup when input with async data', () => {
  //       jest.useFakeTimers();
  //       class AsyncCombobox extends React.Component {
  //         public state = {
  //           data: [],
  //         };
  //         public handleChange = () => {
  //           setTimeout(() => {
  //             this.setState({
  //               data: [{ key: '1', label: '1' }, { key: '2', label: '2' }],
  //             });
  //           }, 500);
  //         };
  //         public render() {
  //           const options = this.state.data.map(item => (
  //             <Option key={item.key}>{item.label}</Option>
  //           ));
  //           return (
  //             <Select
  //               mode="combobox"
  //               onChange={this.handleChange}
  //               filterOption={false}
  //               notFoundContent=""
  //             >
  //               {options}
  //             </Select>
  //           );
  //         }
  //       }
  //       const wrapper = mount(<AsyncCombobox />);
  //       wrapper.find('input').simulate('focus');
  //       wrapper.find('input').simulate('change', { target: { value: '0' } });
  //       jest.runAllTimers();
  //       wrapper.update();
  //       expect(
  //         wrapper
  //           .find('.rc-select-dropdown')
  //           .hostNodes()
  //           .hasClass('rc-select-dropdown-hidden'),
  //       ).toBe(false);
  //     });

  //     // https://github.com/ant-design/ant-design/issues/6600
  //     it('should not repop menu after select', () => {
  //       jest.useFakeTimers();
  //       // tslint:disable-next-line:max-classes-per-file
  //       class AsyncCombobox extends React.Component {
  //         public state = {
  //           data: [{ key: '1', label: '1' }, { key: '2', label: '2' }],
  //         };
  //         public onSelect = () => {
  //           setTimeout(() => {
  //             this.setState({
  //               data: [{ key: '3', label: '3' }, { key: '4', label: '4' }],
  //             });
  //           }, 500);
  //         };
  //         public render() {
  //           const options = this.state.data.map(item => (
  //             <Option key={item.key}>{item.label}</Option>
  //           ));
  //           return (
  //             <Select
  //               mode="combobox"
  //               onSelect={this.onSelect}
  //               filterOption={false}
  //               notFoundContent=""
  //             >
  //               {options}
  //             </Select>
  //           );
  //         }
  //       }
  //       const wrapper = mount(<AsyncCombobox />);
  //       wrapper.find('input').simulate('focus');
  //       wrapper.find('input').simulate('change', { target: { value: '0' } });
  //       expect(
  //         wrapper
  //           .find('.rc-select-dropdown')
  //           .hostNodes()
  //           .hasClass('rc-select-dropdown-hidden'),
  //       ).toBe(false);
  //       wrapper
  //         .find('MenuItem')
  //         .first()
  //         .simulate('click');
  //       jest.runAllTimers();
  //       expect(
  //         wrapper
  //           .find('.rc-select-dropdown')
  //           .hostNodes()
  //           .hasClass('rc-select-dropdown-hidden'),
  //       ).toBe(true);
  //     });
  //   });
  // });

  // it('backfill', () => {
  //   const handleChange = jest.fn();
  //   const handleSelect = jest.fn();
  //   const wrapper = mount<Select>(
  //     <Select
  //       mode="combobox"
  //       backfill={true}
  //       open={true}
  //       onChange={handleChange}
  //       onSelect={handleSelect}
  //     >
  //       <Option value="One">One</Option>
  //       <Option value="Two">Two</Option>
  //     </Select>,
  //   );

  //   const input = wrapper.find('input');

  //   input.simulate('keyDown', { keyCode: KeyCode.DOWN });

  //   expect(wrapper.state().value).toEqual(['Two']);
  //   expect(wrapper.state().backfillValue).toEqual('Two');
  //   expect(wrapper.state().inputValue).toBe('Two');
  //   expect(
  //     wrapper
  //       .find('MenuItem')
  //       .at(1)
  //       .text(),
  //   ).toBe('Two');
  //   expect(handleChange).not.toBeCalled();
  //   expect(handleSelect).not.toBeCalled();

  //   input.simulate('keyDown', { keyCode: KeyCode.ENTER });

  //   expect(wrapper.state().value).toEqual(['Two']);
  //   expect(handleChange).toBeCalledWith('Two', <Option value="Two">Two</Option>);
  //   expect(handleSelect).toBeCalledWith('Two', <Option value="Two">Two</Option>);
  // });

  // it("should hide clear icon when value is ''", () => {
  //   const wrapper = mount(
  //     <Select mode="combobox" value="" allowClear={true}>
  //       <Option value="One">One</Option>
  //       <Option value="Two">Two</Option>
  //     </Select>,
  //   );

  //   expect(wrapper.find('.rc-select-selection__clear').length).toBe(0);
  // });

  // it("should show clear icon when inputValue is not ''", () => {
  //   const wrapper = mount(
  //     <Select mode="combobox" value="One" allowClear={true}>
  //       <Option value="One">One</Option>
  //       <Option value="Two">Two</Option>
  //     </Select>,
  //   );

  //   expect(wrapper.find('.rc-select-selection__clear').length).toBe(1);
  // });

  // it("should hide clear icon when inputValue is ''", () => {
  //   const wrapper = mount(
  //     <Select mode="combobox" allowClear={true}>
  //       <Option value="One">One</Option>
  //       <Option value="Two">Two</Option>
  //     </Select>,
  //   );

  //   wrapper.find('input').simulate('change', { target: { value: '1' } });
  //   expect(wrapper.find('.rc-select-selection__clear').length).toBe(1);
  //   wrapper.find('input').simulate('change', { target: { value: '' } });
  //   expect(wrapper.find('.rc-select-selection__clear').length).toBe(0);
  // });

  // it('autocomplete - option update when input change', () => {
  //   // tslint:disable-next-line:max-classes-per-file
  //   class App extends React.Component {
  //     public state = {
  //       options: [],
  //     };

  //     public updateOptions = value => {
  //       const options = [value, value + value, value + value + value];
  //       this.setState({
  //         options,
  //       });
  //     };

  //     public render() {
  //       return (
  //         <Select mode="combobox" optionLabelProp="children" onChange={this.updateOptions}>
  //           {this.state.options.map(opt => {
  //             return <Option key={opt}>{opt}</Option>;
  //           })}
  //         </Select>
  //       );
  //     }
  //   }

  //   const wrapper = mount(<App />);
  //   wrapper.find('input').simulate('change', { target: { value: 'a' } });
  //   wrapper.find('input').simulate('change', { target: { value: 'ab' } });
  //   expect(wrapper.find('input').prop('value')).toBe('ab');
  //   wrapper
  //     .find('MenuItem')
  //     .at(1)
  //     .simulate('click');
  //   expect(wrapper.find('input').prop('value')).toBe('abab');
  // });

  // // https://github.com/ant-design/ant-design/issues/10367
  // it('should  autocomplete with correct option value', () => {
  //   const wrapper = mount(
  //     <Select mode="combobox" optionLabelProp="children">
  //       {[1, 11, 111, 1111].map(opt => {
  //         return <Option key={opt}>{`xxx-${opt}`}</Option>;
  //       })}
  //     </Select>,
  //   );
  //   wrapper.find('input').simulate('change', { target: { value: '1' } });
  //   wrapper
  //     .find('MenuItem')
  //     .at(0)
  //     .simulate('click');
  //   expect(wrapper.find('input').prop('value')).toBe('xxx-1');
  // });

  // // https://github.com/ant-design/ant-design/issues/16572
  // it('close when enter press without active option', () => {
  //   jest.useFakeTimers();

  //   const onDropdownVisibleChange = jest.fn();

  //   const wrapper = mount<Select>(
  //     <Select mode="combobox" open={true} onDropdownVisibleChange={onDropdownVisibleChange}>
  //       <Option value="One">One</Option>
  //       <Option value="Two">Two</Option>
  //     </Select>,
  //   );

  //   wrapper.find('input').simulate('keyDown', {
  //     keyCode: KeyCode.ENTER,
  //   });

  //   jest.runAllTimers();
  //   wrapper.update();
  //   expect(onDropdownVisibleChange).toBeCalledWith(false);

  //   jest.useRealTimers();
  // });
});
