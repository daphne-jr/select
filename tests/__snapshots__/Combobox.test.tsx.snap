// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.Combobox renders correctly 1`] = `
<div
  class="rc-select rc-select-single"
>
  <div
    class="rc-select-selector"
  >
    <span
      class="rc-select-selection-search"
    >
      <input
        aria-activedescendant="rc_select_6_list_0"
        aria-autocomplete="list"
        aria-controls="rc_select_6_list"
        aria-haspopup="listbox"
        aria-owns="rc_select_6_list"
        autocomplete="off"
        class="rc-select-selection-search-input"
        id="rc_select_6"
        role="combobox"
        value=""
      />
    </span>
    <span
      class="rc-select-selection-placeholder"
    >
      Search
    </span>
  </div>
</div>
`;
